Source/WebCore/rendering/RenderObject.cpp

```diff
@@ -2520,12 +2520,137 @@ Vector<SelectionGeometry> RenderObject::collectSelectionGeometriesWithoutUnionIn
     return collectSelectionGeometriesInternal(range).geometries;
 }
 
+static bool areOnSameLine(const SelectionGeometry& a, const SelectionGeometry& b)
+{
+    if (a.lineNumber() && a.lineNumber() == b.lineNumber())
+        return true;
+
+    auto quadA = a.quad();
+    auto quadB = b.quad();
+    return FloatQuad { quadA.p1(), quadA.p2(), quadB.p2(), quadB.p1() }.isEmpty()
+        && FloatQuad { quadA.p4(), quadA.p3(), quadB.p3(), quadB.p4() }.isEmpty();
+}
+
+static void makeBidiSelectionVisuallyContiguousIfNeeded(const SimpleRange& range, Vector<SelectionGeometry>& geometries)
+{
+    if (!range.startContainer().document().settings().visuallyContiguousBidiTextSelectionEnabled())
+        return;
+
+    FloatPoint selectionStartTop;
+    FloatPoint selectionStartBottom;
+    FloatPoint selectionEndTop;
+    FloatPoint selectionEndBottom;
+
+    std::optional<SelectionGeometry> startGeometry;
+    std::optional<SelectionGeometry> endGeometry;
+    for (size_t i = 0; i < geometries.size(); ++i) {
```

Comment:
can this use a `for const auto& ... : ...` loop instead?

Reply:
Oh, good catch — yes, this should be an iterator for loop since I don't actually need the index (anymore).

Thanks for the review!